src/combine.jl

```diff
@@ -139,3 +139,198 @@ function Base.join{T,N,D,Ax}(As::AxisArray{T,N,D,Ax}...; fillvalue::T=zero(T),
     return result
 
 end #join
+
+function _flatten_array_axes(array_name, array_axes...)
+    ((array_name, (idx isa Tuple ? idx : (idx,))...) for idx in product((Ax.val for Ax in array_axes)...))
+end
+
+function _flatten_axes(array_names, array_axes)
+    collect(Iterators.flatten(map(array_names, array_axes) do tup_name, tup_array_axes
+        _flatten_array_axes(tup_name, tup_array_axes...)
+    end))
+end
+
+function _splitall{N}(::Type{Val{N}}, As...)
+    tuple((Base.IteratorsMD.split(A, Val{N}) for A in As)...)
+end
+
+function _reshapeall{N}(::Type{Val{N}}, As...)
+    tuple((reshape(A, Val{N}) for A in As)...)
+end
+
+function _check_common_axes(common_axis_tuple)
+    if !all(axisname(first(common_axis_tuple)) .=== axisname.(common_axis_tuple[2:end]))
+        throw(ArgumentError("Leading common axes must have the same name in each array"))
+    end
+
+    return nothing
+end
+
+function _flat_axis_eltype(LType, trailing_axes)
+    eltypes = map(trailing_axes) do array_trailing_axes
+        Tuple{LType, eltype.(array_trailing_axes)...}
+    end
+
+    return typejoin(eltypes...)
+end
+
+function collapse{N, AN}(::Type{Val{N}}, As::Vararg{AxisArray, AN})
+    collapse(Val{N}, ntuple(identity, Val{AN}), As...)
+end
+
+function collapse{N, AN, NewArrayType<:AbstractArray}(::Type{Val{N}}, ::Type{NewArrayType}, As::Vararg{AxisArray, AN})
+    collapse(Val{N}, NewArrayType, ntuple(identity, Val{AN}), As...)
+end
+
+@generated function collapse{N, AN, LType}(::Type{Val{N}}, labels::NTuple{AN, LType}, As::Vararg{AxisArray, AN})
+    flat_dim = Val{N + 1}
+    flat_dim_int = Int(N) + 1
+    new_eltype = Base.promote_eltype(As...)
+
+    quote
+        collapse(Val{N}, Array{$new_eltype, $flat_dim_int}, labels, As...)
+    end
+end
+
+"""
+    collapse(::Type{Val{N}}, As::AxisArray...) -> AxisArray
+    collapse(::Type{Val{N}}, labels::Tuple, As::AxisArray...) -> AxisArray
+    collapse(::Type{Val{N}}, ::Type{NewArrayType}, As::AxisArray...) -> AxisArray
+    collapse(::Type{Val{N}}, ::Type{NewArrayType}, labels::Tuple, As::AxisArray...) -> AxisArray
+
+Collapses `AxisArray`s with `N` equal leading axes into a single `AxisArray`.
+All additional axes in any of the arrays are flattened into a single additional
+`CategoricalVector{Tuple}` axis.
+
+### Arguments
+
+* `::Type{Val{N}}`: the greatest common dimension to share between all input
+                    arrays. The remaining axes are flattened. All `N` axes must be common
+                    to each input array, at the same dimension. Values from `0` up to the
+                    minimum number of dimensions across all input arrays are allowed.
+* `labels::Tuple`: (optional) a label for each `AxisArray` in `As` which is used in the flat
```

Comment:
Not entirely clear. Maybe "Index assigned to each array in `As` in the :collapsed axis"? Since it's optional, also specify what happens by default.